flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala

```diff
@@ -0,0 +1,244 @@
+      case EventTimeTumblingGroupWindow(_, _, size) =>
+        createEventTimeTumblingWindowDataSet(inputDS, isTimeInterval(size.resultType))
+      case EventTimeSessionGroupWindow(_, _, _) =>
+        throw new UnsupportedOperationException(
+          "Event-time session windows on batch are currently not supported")
+      case EventTimeSlidingGroupWindow(_, _, _, _) =>
+        throw new UnsupportedOperationException(
+          "Event-time sliding windows on batch are currently not supported")
+      case _: ProcessingTimeGroupWindow =>
+        throw new UnsupportedOperationException(
+          "Processing-time tumbling windows are not supported on batch tables, " +
+            "window on batch must declare a time attribute over which the query is evaluated.")
+    }
+
+    // if the expected type is not a Row, inject a mapper to convert to the expected type
+    expectedType match {
+      case Some(typeInfo) if typeInfo.getTypeClass != classOf[Row] =>
+        val mapName = s"convert: (${getRowType.getFieldNames.toList.mkString(", ")})"
+        result.map(
+          getConversionMapper(
+            config = config,
+            nullableInput = false,
+            inputType = resultRowTypeInfo.asInstanceOf[TypeInformation[Any]],
+            expectedType = expectedType.get,
+            conversionOperatorName = "DataSetWindowAggregateConversion",
+            fieldNames = getRowType.getFieldNames
+          ))
+          .name(mapName)
+      case _ => result
+    }
+  }
+
+
+  private def createEventTimeTumblingWindowDataSet(
+      inputDS: DataSet[Any],
+      isTimeWindow: Boolean)
+    : DataSet[Any] = {
+    val mapFunction = createDataSetWindowPrepareMapFunction(
+      window,
+      namedAggregates,
+      grouping,
+      inputType)
+    val groupReduceFunction = createDataSetWindowAggGroupReduceFunction(
+      window,
+      namedAggregates,
+      inputType,
+      getRowType,
+      grouping,
+      namedProperties)
+
+    val mappedInput = inputDS
+      .map(mapFunction)
+      .name(prepareOperatorName)
+
+    if (isTimeWindow) {
+      // grouped time window aggregation
+      val mapReturnType = mapFunction.asInstanceOf[ResultTypeQueryable[Row]].getProducedType
+      // group by grouping keys and rowtime field (the last field in the row)
+      val groupingKeys = grouping.indices ++ Seq(mapReturnType.getArity - 1)
+      mappedInput.asInstanceOf[DataSet[Row]]
+        .groupBy(groupingKeys: _*)
+        .reduceGroup(groupReduceFunction)
+        .returns(resultRowTypeInfo)
+        .name(aggregateOperatorName)
+        .asInstanceOf[DataSet[Any]]
+    } else {
+      // count window
+      val groupingKeys = grouping.indices.toArray
+      if (groupingKeys.length > 0) {
+        // grouped aggregation
+        mappedInput.asInstanceOf[DataSet[Row]]
+          .groupBy(groupingKeys: _*)
+          // sort on time field, it's the one after grouping keys
+          .sortGroup(groupingKeys.length, Order.ASCENDING)
```

Comment:
Shouldn't this be `mapReturnType.getArity - 1`? According to the docs of `AggregateUtil#createDataSetWindowPrepareMapFunction` the time field should be at the end?